docs/rfcs/9999-vpn-firefox-integration.md

```diff
@@ -0,0 +1,188 @@
+
+## Delivery
+As the Extension will be Mozilla signed, asking Firefox to open a URL pointing to our XPI will prompt the install, without warnings. 
+The onboarding will be refined based on foxfooding feedback. 
+
+
+## VPN API 
+
+
+### Networking
+Using the [proxy api](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/proxy) a network extension can decide which proxy each network request should use. So from a network request perspective, there are only 4 possible cases to implement:
+
+| StateNr               | Device VPN is On | Device VPN is OFF |
+|-----------------|------------------|-------------------|
+| Page VPN is On  | 1                | 2                 |
+| Page VPN is Off | 3                | 4                 |
+
+In the following sections, we will explain what changes we need inside the client to support each of the cases and where the extension must proxy traffic to. 
+
+
+#### State 1: Device VPN is On & Page VPN is On
+In this state the extension can simply set the proxy for those network requests to `null` to use the user's active server.  
+If a custom exit location is set it must set it to the `socks5` proxy of that endpoint. No changes are needed on the client as this is how MAC uses the current API.
+
+
+#### State 2: Device VPN is OFF & Page VPN is On
+
+Simplified, the VPN has the following states:
+```cpp
+  enum State {
+    StateOff, // Nothing protected
+    StateOn, // Whole device protected
+    (...) // unrelated intermediate states
+  };
+```
+An extension can currently send an `ExtensionMessage<activate>` to the VPN to turn on whole-device protection. 
+For this case, we propose to introduce a new state `StateOnPartial` that will be activated on a call of an extension to `ExtensionMessage<activate>` instead. It is similar to `StateOn` with the following properties: 
+- The `allowedIPAddressRanges` for this activation is set to `10.124.0.0/20` so that only the Mullvad Exit Proxies are accessible on the VPN Network Device. Therefore, all other traffic will not use the VPN. 
+- A Mullvad DNS server will be set, so that the Mullvad proxies are resolvable. If a custom DNS is set, we need to set a Mullvad DNS as 2nd DNS server.
+- The Client UI does not show an activation. (or a new special UI state)
+> Q: Do we need a new UI state ? or like a "firefox is using the vpn?" 
+- If the Client activates, it will attempt to switch to `StateOn` and protect the whole device.
+- An Extension may request `ExtensionMessage<deactivate>` only if the client is in `StateOnPartial`. The Client will then go to `StateOff`.
+
+A simplified graph excluding the intermediate states. 
+
+```mermaid 
+---
+title: State Diagramm v2
+---
+stateDiagram-v2
+    StateOff --> StateOn: activate
+    StateOff --> StateOnPartial: activatePartial
+    StateOnPartial --> StateOn: silentServerSwitch
+    StateOnPartial --> StateOff: deactivate
+    StateOn -->StateOff : deactivate 

```

Comment:
> StateOn -->StateOff : deactivate

It appears that deactivating the Client in StateOn will turn off the VPN for a Firefox tab. Is that what we want? Or should we transition to StateOnPartial in that case?